File: 725 Split Linked List in Parts/solution_2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode[] splitListToParts(ListNode head, int k) {

        int cnt = 0;
        ListNode ptr = head;
        
        while (head != null)
        {
            cnt++;
            head = head.next; 
        }

        int[] cnts = new int[k];
        int per_cnt = cnt / k;
        int rem = cnt % k;

        for (int i=0;i<k;i++)
        {
            cnts[i] = per_cnt + (rem>0 ? 1 : 0);
            if (rem>0)
                rem--;
        }

        ListNode[] ans = new ListNode[k];

        for (int i=0;i<k;i++)
        {
            ListNode currHead = ptr;
            while (cnts[i]-->0)
            {
                if (cnts[i] == 0)
                {
                    ListNode next_ = ptr.next;
                    ptr.next = null;
                    ptr = next_;
                }
                else
                {
                    ptr = ptr.next;
                }
            }

            ans[i] = currHead;
        }
        
        return ans;
    }
}

File: 2807 Insert Greatest Common Divisors in Linked List/solution_2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* insertGreatestCommonDivisors(ListNode* head) {
        
        ListNode *p1 = head, *p2 = head->next;

        while (p2)
        {
            ListNode *gcd = new ListNode(__gcd(p1->val, p2->val), p2);
            p1->next = gcd;
            p1 = p1->next->next;
            p2 = p2->next;
        }

        return head;
    }
};
